pctasks/run/pctasks/run/task/batch.py

```diff
@@ -213,60 +240,121 @@ def get_failed_tasks(
+                    error: Any = e.error  # Avoid type hinting error
+                    if error.code == "JobNotFound":
+                        job_failed = True
+                    else:
+                        logger.exception(
+                            f"(BATCH) Error getting job info for {job_id}."
+                        )
+
+                logger.info(
+                    "(BATCH) Polling task runner for "
+                    f"failed tasks in job {job_id}..."
+                )
+                try:
+                    failed_tasks = batch_client.get_failed_tasks(job_id)
+                except Exception:
+                    logger.exception("(BATCH) Error getting failed tasks.")
+                    # Don't let a Batch API error fail the job;
+                    # consider it as "no reported failures"
+                    failed_tasks = {}
+
+                total_failed = [len(failed_tasks)]
+                if total_failed:
+                    logger.info(f"(BATCH)  - {sum(total_failed)} failed tasks found.")
+                else:
+                    logger.info("(BATCH)  - No failed tasks found.")
+
+                result = (failed_tasks, job_failed)
+                self.response_cache[cache_key] = result
+                return result
+
     def poll_task(
         self,
         runner_id: Dict[str, Any],
         previous_poll_count: int,
     ) -> TaskPollResult:
+        batch_client = self._get_batch_client()
+
         task_id = BatchTaskId.parse_obj(runner_id)
-        with BatchClient(self.settings.batch_settings) as batch_client:
-            task_status_result = batch_client.get_task_status(
-                job_id=task_id.batch_job_id, task_id=task_id.batch_task_id
-            )
 
-            if task_status_result is None:
-                if previous_poll_count < MAX_MISSING_POLLS:
-                    return TaskPollResult(task_status=TaskRunStatus.PENDING)
-                else:
-                    return TaskPollResult(
-                        task_status=TaskRunStatus.FAILED,
-                        poll_errors=[
-                            f"Batch task not found after {previous_poll_count} polls."
-                        ],
-                    )
+        task_status_result = batch_client.get_task_status(
+            job_id=task_id.batch_job_id, task_id=task_id.batch_task_id
+        )
+
+        if task_status_result is None:
+            if previous_poll_count < MAX_MISSING_POLLS:
+                return TaskPollResult(task_status=TaskRunStatus.PENDING)
             else:
-                task_status, error_message = task_status_result
                 return TaskPollResult(
-                    task_status=task_status,
-                    poll_errors=map_opt(lambda e: [e], error_message),
+                    task_status=TaskRunStatus.FAILED,
+                    poll_errors=[
+                        f"Batch task not found after {previous_poll_count} polls."
```

Comment:
I wasn't aware of this `MAX_MISSING_POLLS` concept.

We've been seeing issues where the Jobs were "failed" (and so the Batch Tasks cancelled) when the orchestrator had issues getting a status from Azure Batch. With this PR, getting the task statuses should be much more reliable. But it still feels wrong to cancel a (perhaps perfectly healthy) Batch Task just because the orchestrator couldn't check in on it.

What's the failure mode if the orchestrator really never hears back from Batch. Then the job is indefinitely in PENDING / RUNNING state? So we're taking up some resources on the orchestrator node, and the UI might show a PENDING / RUNNING job, despite it actually being finished in Batch. To me, that outcome sounds better than cancelling a running batch task.

Reply:
Currently, if the orchestrator doesn't hear from Batch polls for a task 5 times (spaced out through time based on the poll frequency, which defaults to 30 seconds), then it will fail the task, which would eventually fail the job and workflow. If there's an issue submitting the job that is not immediately reported as a Batch API error, then this is the mechanism that ensures the orchestrator isn't sitting waiting for tasks that will never materialize.

I can't remember the exact situation, but I'm pretty sure I put this in to work around a specific failure with the task never materializing in Batch. Have we seen this error show up in practice, where 2.5 minutes pass without the task being able to report status and it not being a failure mode, so the task gets cancelled? 

Maybe we can raise the default from 5 to give a bigger runway? Or we could remove it and see if we experience the dangling orchestrator jobs in other failure modes. I'm happy to go either direction.

Reply:
> I can't remember the exact situation, but I'm pretty sure I put this in to work around a specific failure with the task never materializing in Batch. Have we seen this error show up in practice, where 2.5 minutes pass without the task being able to report status and it not being a failure mode, so the task gets cancelled?

I believe so, but I might be wrong. `pctasks runs get workflow 4d3502e5-a476-4ecb-8971-c87053bc8bc6` against staging is one I believe.


```
- type: JobRun
  schema_version: 1.0.0
  deleted: false
  status: failed
  status_history:
  - status: pending
    timestamp: '2023-02-24T12:31:20.560813+00:00'
  - status: running
    timestamp: '2023-02-24T12:36:06.320130+00:00'
  - status: failed
    timestamp: '2023-02-24T15:00:33.307561+00:00'
  run_id: 4d3502e5-a476-4ecb-8971-c87053bc8bc6
  job_id: process-chunk
  job_partition_counts:
    pending: 0
    running: 31
    completed: 5
    failed: 0
    cancelled: 0
  errors:
  - 'Job partitions thread failed with Error occurred in request., ConnectTimeout:
    HTTPSConnectionPool(host=''pc7daq7staging7etl7batch.westeurope.batch.azure.com'',
    port=443): Max retries exceeded with url: /jobs/goes_glm-process-chunk-4d3502e5--cb-8971-c87053bc8bc6-tsk_gen_bl/tasks?api-version=2021-06-01.14.0&$filter=state%20eq%20%27completed%27&maxresults=1000&timeout=30
    (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x7f901c5c2880>,
    ''Connection to pc7daq7staging7etl7batch.westeurope.batch.azure.com timed out.
    (connect timeout=100)''))'
  - 'Job partitions thread failed with Error occurred in request., ConnectTimeout:
    HTTPSConnectionPool(host=''pc7daq7staging7etl7batch.westeurope.batch.azure.com'',
    port=443): Max retries exceeded with url: /jobs/goes_glm-process-chunk-4d3502e5--cb-8971-c87053bc8bc6-tsk_gen_bl/tasks?api-version=2021-06-01.14.0&$filter=state%20eq%20%27completed%27&maxresults=1000&timeout=30
    (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x7f901efd4e80>,
    ''Connection to pc7daq7staging7etl7batch.westeurope.batch.azure.com timed out.
    (connect timeout=100)''))'
```

That *specific* failure is hopefully resolved by this PR, by not overwhelming the Batch API. But I think the same general concern applies.

> I'm happy to go either direction.

I'm also happy either way. Let's leave it as is (fail a task we don't hear about for a while), and revisit if things if it's a problem. At the very least, failing will force us to fix issues like this one where we were overwhelming the Batch API.